store/postgres/src/deployment_store.rs

```diff
@@ -1170,29 +1170,20 @@ impl DeploymentStore {
         Ok(())
     }
 
-    /// This should be called once per subgraph on `graph-node` initialization,
-    /// before processing the first block on start.
-    ///
-    /// It does nothing if there's no fatal error for a subgraph, or the blocks received are `None`.
-    /// It branches behavior based of if the fatal error is deterministic or not, read the specific
-    /// functions down below.
-    pub(crate) fn unfail(
-        &self,
-        site: Arc<Site>,
-        current_ptr: Option<&BlockPtr>,
-        parent_ptr: Option<&BlockPtr>,
-    ) -> Result<(), StoreError> {
-        let current_ptr = match current_ptr {
-            Some(current_ptr) => current_ptr,
-            // No current head pointer, nothing to do
-            None => return Ok(()),
-        };
-
+    // Base for unfail logic, basically it early returns / does nothing if:
+    //
+    // - There's no fatal error for a subgraph
+    // - The error's determinism isn't the same as the one requested
+    //
+    // Otherwise it calls the provided closure/function with the subgraph error.
+    fn unfail_error<F>(&self, site: &Arc<Site>, deterministic: bool, f: F) -> Result<(), StoreError>
+    where
+        F: Fn(&PgConnection, &DeploymentHash, &ErrorDetail) -> Result<(), StoreError>,
```

Comment:
I dislike this callback style, I'd rather duplicate the function for the deterministic and non-deterministic cases, and then factor out code that can be reused between them.

Reply:
Do you have a suggestion for doing this without the callback style considering that we need early returns here?

I could do a common method, but that would have to return: `should_early_return: bool` + `maybe_subgraph_error: Option<ErrorDetail>` values, which I dislike more over the callback style.

Example:

```rust
fn unfail_error() -> (bool, Option<ErrorDetail>) {

}

fn unfail_deterministic() {
    let (should_early_return, maybe_subgraph_error) = unfail_error();

    if should_early_return {
        return Ok(());
    }

    let subgraph_error = maybe_subgraph_error.unwrap();

    // ... main logic
}

fn unfail_non_deterministic() {
    let (should_early_return, maybe_subgraph_error) = unfail_error();

    if should_early_return {
        return Ok(());
    }

    let subgraph_error = maybe_subgraph_error.unwrap();

    // ... main logic
}
```

Reply:
IMO when we're talking about having to control flow like this (early returning), the callback style is more suitable.

And in the future if these two methods don't have anything in common anymore, the developer coding over this should remove the common function.

Reply:
Abstraction removed, and `error::detail` one improved.

Fixed.